ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,271 @@
+from ...ext import SpanTypes
+
+
+if TYPE_CHECKING:
+    from typing import Callable
+    from typing import Dict
+    from typing import Iterable
+    from typing import List
+    from typing import Tuple
+    from typing import Union
+
+    from ddtrace import Span
+
+
+_graphql_version_str = getattr(graphql, "__version__")
+graphql_version = parse_version(_graphql_version_str)
+
+if graphql_version < (3, 0):
+    from graphql.language.ast import Document
+else:
+    from graphql.language.ast import DocumentNode as Document
+
+config._add("graphql", dict(_default_service="graphql"))
+
+
+def _graphql_resolvers_enabled():
+    """
+    Patching graphql resolvers may produce large traces.
+    DD_TRACE_GRAPHQL_RESOLVERS_ENABLED is disabled by default.
+    """
+    return asbool(os.getenv("DD_TRACE_GRAPHQL_RESOLVERS_ENABLED", default=False))
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False):
+        return
+    setattr(graphql, "_datadog_patch", True)
+    Pin().onto(graphql)
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(wrap, module_str, func_name, wrapper)
+
+
+def unpatch():
+    if not getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+
+    for module_str, func_name, wrapper in _get_patching_candidates():
+        _update_patching(unwrap, module_str, func_name, wrapper)
+
+    setattr(graphql, "_datadog_patch", False)
+
+
+def _get_patching_candidates():
+    if graphql_version < (3, 0):
+        return [
+            ("graphql.graphql", "execute_graphql", _traced_query),
+            ("graphql.language.parser", "parse", _traced_parse),
+            ("graphql.validation.validation", "validate", _traced_validate),
+            ("graphql.execution.executor", "execute", _traced_execute),
+        ]
+    return [
+        ("graphql.graphql", "graphql_impl", _traced_query),
+        ("graphql.language.parser", "parse", _traced_parse),
+        ("graphql.validation.validate", "validate", _traced_validate),
+        ("graphql.execution.execute", "execute", _traced_execute),
+    ]
+
+
+def _update_patching(operation, module_str, func_name, wrapper):
+    module = sys.modules[module_str]
+    func = getattr(module, func_name)
+    operation(func, wrapper)
+
+
+def _traced_parse(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    with pin.tracer.trace(
+        name="graphql.parse",
+        service=trace_utils.int_service(pin, config.graphql),
+        span_type=SpanTypes.GRAPHQL,
+    ) as span:
+        _init_span(span)
+        return func(*args, **kwargs)
+
+
+def _traced_validate(func, args, kwargs):
+    pin = Pin.get_from(graphql)
+    if not pin or not pin.enabled():
+        return func(*args, **kwargs)
+
+    with pin.tracer.trace(
+        name="graphql.validate",
+        service=trace_utils.int_service(pin, config.graphql),
```

Comment:
do we need to get setting service explicitly on all of these, won't most inherit the proper service name from it's parent?

Reply:
In the majority of graphql libraries `graphql.parse()`, `graphql.validate()`, and `graphql.execute()` will be called in `graphql.graphql()`. In this case `graphql.query` will be the only top level span and do not need to set the service explicitly. However if these methods are called directly then these methods will be top level. 

Here's an example where graphql.parse is a top level span:
https://github.com/DataDog/dd-trace-py/blob/a73db8f4f272b56786c99f2dde3c9361680e0699/tests/snapshots/tests.contrib.graphql.test_graphql.test_graphql_v2_with_document.json